internal/elasticsearch/cluster/script.go

```diff
@@ -0,0 +1,151 @@
+package cluster
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/elastic/terraform-provider-elasticstack/internal/clients"
+	"github.com/elastic/terraform-provider-elasticstack/internal/models"
+	"github.com/elastic/terraform-provider-elasticstack/internal/utils"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+func ResourceScript() *schema.Resource {
+	scriptSchema := map[string]*schema.Schema{
+		"script_id": {
+			Description: "Identifier for the stored script. Must be unique within the cluster.",
+			Type:        schema.TypeString,
+			Required:    true,
+			ForceNew:    true,
+		},
+		"lang": {
+			Description:  "Script language. For search templates, use `mustache`. Defaults to `painless`.",
+			Type:         schema.TypeString,
+			Optional:     true,
+			Default:      "painless",
+			ValidateFunc: validation.StringInSlice([]string{"painless", "expression", "mustache", "java"}, false),
+		},
+		"source": {
+			Description: "For scripts, a string containing the script. For search templates, an object containing the search template.",
+			Type:        schema.TypeString,
+			Required:    true,
+		},
+		"params": {
+			Description:      "Parameters for the script or search template.",
+			Type:             schema.TypeString,
+			Optional:         true,
+			DiffSuppressFunc: utils.DiffJsonSuppress,
+			ValidateFunc:     validation.StringIsJSON,
+		},
+		"context": {
+			Description: "Context in which the script or search template should run.",
+			Type:        schema.TypeString,
+			Optional:    true,
+		},
+	}
+	utils.AddConnectionSchema(scriptSchema)
+
+	return &schema.Resource{
+		Description: "Creates or updates a stored script or search template. See https://www.elastic.co/guide/en/elasticsearch/reference/current/create-stored-script-api.html",
+
+		CreateContext: resourceScriptPut,
+		UpdateContext: resourceScriptPut,
+		ReadContext:   resourceScriptRead,
+		DeleteContext: resourceScriptDelete,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: scriptSchema,
+	}
+}
+
+func resourceScriptRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var diags diag.Diagnostics
+	client, err := clients.NewApiClient(d, meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	id := d.Id()
+	compId, diags := clients.CompositeIdFromStr(id)
+	if diags.HasError() {
+		return diags
+	}
+
+	script, diags := client.GetElasticsearchScript(ctx, compId.ResourceId)
+	if !d.IsNewResource() && script == nil && diags == nil {
+		tflog.Warn(ctx, fmt.Sprintf(`Script "%s" not found, removing from state`, compId.ResourceId))
```

Comment:
It doesn't look like we log a message in this case for the other resources. Any reason this should be a warning?

Reply:
Yes, the reason is to let the user know that the resource is removed from state when actually not found on API.
I think it's pretty common pattern which the other providers do too.
https://github.com/hashicorp/terraform-provider-aws/search?q=removing+from+state

It it makes sense, I'll add it to the other resources with `!d.IsNewResource()` check.
https://github.com/hashicorp/terraform-provider-aws/issues/16796

Reply:
I just realized that adding `!d.IsNewResource()` is not good here since it can cause null pointer error in the following code.
It might be better to return error when the resource is just created resource as aws provider does, but I simply removed to align with the other resources for now.